### ISSUE TITLE
Filter strings that look like pronunciations, but aren't

### ISSUE THREAD

**Issue by DanielSWolf** [tatuylonen/wiktextract#48]
The extracted JSON data contains many `sounds.ipa` entries that aren't actually meant to be pronunciations of the current word. It seems to me that most of them are caused by one of the following two problems:

### 1. The IPAchar macro

The *IPAchar* macro is used in Wiktionary to indicate that the content is not meant to be a pronunciation of the current word, but is merely used to illustrate some aspect of its pronunciation. However, Wiktextract treats it as a pronunciation. See the following examples.

| Word | Detected pronunciation | Relevant Wikitext code |
|---|---|---|
| Fussel | `{"ipa": "/s/"}` | {{q\|common variant; <u>not</u> restricted to accents with general voicing of {{IPAchar\|/s/}}}} |
| Kokos | `{"ipa": "/\ɔ/"}`, `{"ipa": "[o]"}` | {{q\|entirely uncommon in northern and central Germany, perhaps used by southern speakers whose {{IPAchar\|/ɔ/}} is {{IPAchar\|[o]}}}} |

The English Wiktionary contains 23,665 uses of *IPAchar*, leading to a large number of false pronunciations. It would be great if *IPAchar* macros could be filtered from the pronunciation list. 

### 2. Audio transcripts

An audio recording often contains more than just the current word. For instance, it may contain multiple pronunciation variations of the same word, or it may prefix a noun with an article. If an audio file and a pronunciation are on the same line of Wikitext, this typically indicates that the pronunciation is a phonetic transcript of the recording, not a pronunciation of the current word. See the following examples.

| Word | Detected pronunciation | Relevant Wikitext code |
|---|---|---|
| Kaffee | `{"ipa": "[kʰaˈfeː‿ˈkʰafe]"}` | * {{audio\|de\|De-Kaffee.ogg}} {{IPA\|de\|[kʰaˈfeː‿ˈkʰafe]}} |
| Herr | `{"ipa": "[dɜɾ‿hɛr]", "tags": ["Austria"]}` | {{a\|Austria}} {{audio\|de\|De-at-Herr.ogg}} {{IPA\|de\|[dɜɾ‿hɛr]}} {{q\|[[der]] Herr}}
| taxi | `{"ipa": "[un ˈtaksi]"}` | {{audio-IPA\|it\|It-un taxi.ogg\|[un ˈtaksi]}} |

Note that there are two different syntaxes: The first two examples combine *audio* and *IPA*, while the third example uses *audio-IPA*.

The English Wiktionary edition contains 92 cases of *audio*+*IPA* and 158 cases of *audio-IPA*. It might make sense to exclude these from the pronunciation lists, but I don't think this case is as clear-cut as with the *IPAchar* macro.

**Comment by tatuylonen** [comment]
Thanks for the detailed report!

The IPAchar issue is a bit tricky one.  It doesn't look like I can just suppress the IPAchar template globally (which would be easy to do with two lines of code), as several language-specific IPA templates expand to it and even some of their Lua module implementations emit {{IPAchar|...}} templates.

It also looks like simply stripping all top-level IPAchar calls may also not be correct.  For example, the following entries use IPAchar for what looks like an actual intended alternative IPA: quahog/English, meck/Swedish, Roosevelt/English, par/Gabrielino-Fernandeño, pall mall/English, paths/English, patriota/Polish, pasivum/Czech (there are many others).  With a quick search I'm finding 109 instances of {{IPA|...}} or {{<something>-IPA...}} followed on the same line by ", {{IPAchar|", and in most of them the IPAchar looks like a valid pronunciation that should be collected.  Many of these have a <ref> tag following each translation but many also just use it as an alternative to having multiple pronunciations in a single IPA template.

I'm actually wondering if stripping the "qualifier" and "q" templates might help, together with stripping <ref> tags.  On the other hand, given the seemingly low number of uses of IPAchar that need to be collected, one option would also be to just fix the known ones and then strip top-level IPAchar templates.  However I don't know how <ref> should be used with IPA.

Any opinions?

As for the audio transcript IPAs, those look like a fairly straightforward case, at least for the listed special cases.  I may need to restructure the code somewhat to handle the individual items properly (it is in need of that restructuring anyway).

I'm planning to do some improvements to translations parsing first.  I'm hoping to get to these pronunciation fixes next week.

**Comment by DanielSWolf** [comment]
*Note: In your sentence "However I don't know how \<ref\> should be used with IPA.", the word  \<ref\> isn't rendered because it looks like HTML.*

I see your point about *IPAchar* being used for regular pronunciations. It looks like the issues I'm having are all caused by *IPAchar* templates within *q* templates, so maybe stripping those could be enough for now.

BTW, I admire your commitment to this project! Extracting structured data from Wikitext is certainly no fun. I have a recurring Phantasy of an alternative world where Wiktionary is based on a unified, structured database for all editions. 😁

**Comment by DanielSWolf** [comment]
I ran into a number of other false positives. I fixed most of them on Wiktionary, but I'm not sure about this one:

The *Usage notes* section of [aquatic](https://en.wiktionary.org/wiki/aquatic) uses `{{IPA|en|/w/}}`, which gets picked up by Wiktextract. This is a context where I don't think that `{{q}}` would make much sense. Can you recommend a syntax that won't get picked up?

**Comment by DanielSWolf** [comment]
Here is another case that might require changes to the parsing logic:

| Word | Detected pronunciation | Relevant Wikitext code |
|---|---|---|
| [can](https://en.wiktionary.org/wiki/can#Pronunciation) | `/æ/` | `{{a\|see [[w:/æ/ raising]]}}` |

**Comment by tatuylonen** [comment]
All of these issues should now be fixed:
  - it now ignores content inside "q", "a", and a few other templates when parsing IPAs
  - it now doesn't capture IPAs from lines with audio files into "ipas", but instead puts it into "audio-ipa" with the audio file
  - it no longer captures pronunciations from subsections of the Pronunciation section (this handles the Usage notes section)

The changes should be reflected on https://kaikki.org/ by now.

**Comment by DanielSWolf** [comment]
Thank you! It took me some time to test because I was having network problems, but the pronunciations are much better now!

**Comment by tatuylonen** [comment]
I've noticed that there are some pronunciations that start with a dash, and apparently only provide a suffix for some alternative pronunciations.  These are in Wiktionary.  I'm not quite sure what to do about them.  I doubt they can be reliably expanded to full pronunciations, but I haven't really looked into the matter.

**Comment by DanielSWolf** [comment]
I know. The same is true for pronunciations *ending* with a dash, which are only prefixes.

In my project, I'm discarding any pronunciations starting or ending with a dash because I couldn't find a reliable way of expanding them, either.

This is a slippery slope, however: I'm using a whole bunch of heuristics to filter out bad pronunciations. And don't even get me started on normalization, like people using 'g' (Latin small letter g) instead of 'ɡ' (Latin small letter script g) or ':' instead of 'ː'. Unless you want to go down that path, it might make sense to make sure that your data faithfully reflects the sources, but to leave further processing to downstream projects.

**Comment by tatuylonen** [comment]
If you are interested in sharing your IPA canonicalization code, I would be interested in integrating it.  I think we should still make the original IPAs available, but I think it would help a number of people if a canonicalized version was readily available (assuming the canonicalization is relatively reliable).

**Comment by DanielSWolf** [comment]
A few days ago, I published the first version of [WikiPronunciationDict](https://github.com/DanielSWolf/wiki-pronunciation-dict), a multilingual pronunciation dictionary based on several editions of Wiktionary. For the English Wiktionary edition, I'm using your wiktextract file as source; for the other editions (which tend to use far less Lua), I directly parse the dumps. (This is what I had in mind when I mentioned downstream projects.)

WikiPronunciationDict goes to some lengths to ensure optimal normalization of pronunciations. It uses a hand-rolled IPA parser that transforms all the symbols, diacritics, and suprasegmentals into a structured representation. The parser has special rules for uncommon and language-specific symbols and for common mistakes when authoring IPA strings. After parsing, the structured data is used to re-assemble an IPA string.

Clearly, this is not what you want to do for wiktextract. But I'm afraid that "a bit of normalization" may be worse than none at all. If a program knows it's dealing with unprocessed user data, it will be wary of all the possible errors that may be contained. For instance, some Wiktionary pronunciation strings really contain plain English text rather than a valid pronunciation. A program handling this original input can use some heuristic to determine that this is not, in fact, an IPA string. If, on the other hand, some normalization has already taken place, this non-IPA text may now look much more like a plausible IPA string.

If you're really determined to take this path, I can provide you with some replacement tables with common errors and their fixes. What's important to me, though, is that you keep providing the raw pronunciation strings as well.
